File: cli/internal/scanner/scanner.go
package scanner

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/m2arm/cli/internal/sdk"
)

// Scanner handles code scanning for x86-specific patterns
type Scanner struct {
	projectPath string
	config      map[string]interface{}
	
	// Pattern definitions for x86-specific code
	patterns map[string][]string
	
	// File extensions to scan
	extensions map[string]bool
}

// New creates a new scanner instance
func New(projectPath string, config map[string]interface{}) *Scanner {
	scanner := &Scanner{
		projectPath: projectPath,
		config:      config,
		patterns: map[string][]string{
			"inline_assembly": {
				`__asm__\s*\(`,
				`asm\s*\(`,
				`_asm\s*{`,
			},
			"x86_intrinsics": {
				`#include\s*<.*mmintrin\.h.*>`,
				`#include\s*<.*xmmintrin\.h.*>`,
				`#include\s*<.*emmintrin\.h.*>`,
				`#include\s*<.*pmmintrin\.h.*>`,
				`#include\s*<.*immintrin\.h.*>`,
				`_mm_\w+`,
				`_mm\d+_\w+`,
			},
			"architecture_checks": {
				`#ifdef\s+_M_X64`,
				`#ifdef\s+__x86_64__`,
				`#ifdef\s+_M_IX86`,
				`#ifdef\s+__i386__`,
			},
			"platform_specific": {
				`GetSystemInfo`,
				`IsWow64Process`,
				`SYSTEM_INFO`,
			},
		},
		extensions: map[string]bool{
			".c":    true,
			".cpp":  true,
			".cc":   true,
			".cxx":  true,
			".h":    true,
			".hpp":  true,
			".hxx":  true,
			".py":   true,
			".go":   true,
			".rs":   true,
			".java": true,
			".cs":   true,
			".js":   true,
			".ts":   true,
			".jsx":  true,
			".tsx":  true,
		},
	}

	return scanner
}

// Scan performs the code scanning
func (s *Scanner) Scan() (*sdk.ScanResults, error) {
	results := &sdk.ScanResults{
		ScanTime: time.Now(),
	}

	// Get list of files to scan
	files, err := s.getSourceFiles()
	if err != nil {
		return nil, fmt.Errorf("failed to get source files: %w", err)
	}

	results.TotalFiles = len(files)

	// Scan each file
	for _, file := range files {
		issues, err := s.scanFile(file)
		if err != nil {
			if s.isVerbose() {
				fmt.Fprintf(os.Stderr, "Warning: failed to scan %s: %v\n", file, err)
			}
			continue
		}

		results.Issues = append(results.Issues, issues...)
		results.ScannedFiles++
	}

	// Scan build systems
	buildSystems, err := s.scanBuildSystems()
	if err != nil {
		return nil, fmt.Errorf("failed to scan build systems: %w", err)
	}
	results.BuildSystems = buildSystems

	// Scan dependencies
	dependencies, err := s.scanDependencies()
	if err != nil {
		return nil, fmt.Errorf("failed to scan dependencies: %w", err)
	}
	results.Dependencies = dependencies

	// Generate recommendations
	results.Recommendations = s.generateRecommendations(results)

	return results, nil
}

// getSourceFiles returns a list of source files to scan
func (s *Scanner) getSourceFiles() ([]string, error) {
	var files []string

	err := filepath.Walk(s.projectPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip directories
		if info.IsDir() {
			// Skip common directories that shouldn't be scanned
			name := info.Name()
			if name == ".git" || name == "__pycache__" || name == "node_modules" ||
				name == ".venv" || name == "venv" || name == "build" || name == "dist" ||
				name == ".tox" || name == ".pytest_cache" {
				return filepath.SkipDir
			}
			return nil
		}

		// Check if file extension should be scanned
		ext := strings.ToLower(filepath.Ext(path))
		if s.extensions[ext] {
			files = append(files, path)
		}

		return nil
	})

	return files, err
}

// scanFile scans a single file for x86-specific patterns
func (s *Scanner) scanFile(filePath string) ([]sdk.Issue, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var issues []sdk.Issue
	scanner := bufio.NewScanner(file)
	lineNum := 0

	for scanner.Scan() {
		lineNum++
		line := scanner.Text()

		// Check each category of patterns
		for category, patterns := range s.patterns {
			for _, pattern := range patterns {
				regex, err := regexp.Compile(pattern)
				if err != nil {
					continue // Skip invalid regex
				}

				if match := regex.FindString(line); match != "" {
					relPath, _ := filepath.Rel(s.projectPath, filePath)
					
					issue := sdk.Issue{
						File:        relPath,
						Line:        lineNum,
						Category:    category,
						Pattern:     pattern,
						MatchedText: match,
						Severity:    s.getSeverity(category),
						Suggestion:  s.getSuggestion(category, match),
					}
					issues = append(issues, issue)
				}
			}
		}
	}

	return issues, scanner.Err()
}

// getSeverity returns the severity level for an issue category
func (s *Scanner) getSeverity(category string) string {
	severityMap := map[string]string{
		"inline_assembly":     "high",
		"x86_intrinsics":     "high",
		"architecture_checks": "medium",
		"platform_specific":   "medium",
	}
	if severity, ok := severityMap[category]; ok {
		return severity
	}
	return "low"
}

// getSuggestion returns a suggestion for fixing an issue
func (s *Scanner) getSuggestion(category, matchedText string) string {
	suggestions := map[string]string{
		"inline_assembly":     "Replace with portable C/C++ code or use ARM NEON intrinsics",
		"x86_intrinsics":     "Replace with ARM NEON equivalents or portable alternatives",
		"architecture_checks": "Add ARM architecture checks or use runtime detection",
		"platform_specific":   "Use cross-platform alternatives or add ARM-specific implementations",
	}
	if suggestion, ok := suggestions[category]; ok {
		return suggestion
	}
	return "Review for ARM compatibility"
}

// scanBuildSystems scans for build system configurations
func (s *Scanner) scanBuildSystems() ([]sdk.BuildSystem, error) {
	var buildSystems []sdk.BuildSystem

	buildPatterns := map[string]string{
		"CMakeLists.txt": "cmake",
		"Makefile":       "make",
		"build.gradle":   "gradle",
		"pom.xml":        "maven",
		"package.json":   "npm",
		"Cargo.toml":     "cargo",
		"go.mod":         "go_modules",
	}

	for pattern, system := range buildPatterns {
		matches, err := filepath.Glob(filepath.Join(s.projectPath, pattern))
		if err != nil {
			continue
		}

		for _, match := range matches {
			relPath, _ := filepath.Rel(s.projectPath, match)
			buildSystems = append(buildSystems, sdk.BuildSystem{
				File:        relPath,
				System:      system,
				NeedsReview: true,
			})
		}

		// Also check subdirectories
		err = filepath.Walk(s.projectPath, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}

			if !info.IsDir() && info.Name() == pattern {
				relPath, _ := filepath.Rel(s.projectPath, path)
				buildSystems = append(buildSystems, sdk.BuildSystem{
					File:        relPath,
					System:      system,
					NeedsReview: true,
				})
			}

			return nil
		})
	}

	return buildSystems, nil
}

// scanDependencies scans project dependencies
func (s *Scanner) scanDependencies() ([]sdk.Dependency, error) {
	var dependencies []sdk.Dependency

	// Check package.json (Node.js)
	if deps, err := s.scanNpmDependencies(); err == nil {
		dependencies = append(dependencies, deps...)
	}

	// Check requirements.txt (Python)
	if deps, err := s.scanPythonDependencies(); err == nil {
		dependencies = append(dependencies, deps...)
	}

	// Check Cargo.toml (Rust)
	if deps, err := s.scanCargoDependencies(); err == nil {
		dependencies = append(dependencies, deps...)
	}

	// Check go.mod (Go)
	if deps, err := s.scanGoDependencies(); err == nil {
		dependencies = append(dependencies, deps...)
	}

	return dependencies, nil
}

// scanNpmDependencies scans npm dependencies from package.json
func (s *Scanner) scanNpmDependencies() ([]sdk.Dependency, error) {
	packageJsonPath := filepath.Join(s.projectPath, "package.json")
	if _, err := os.Stat(packageJsonPath); os.IsNotExist(err) {
		return nil, err
	}

	// For simplicity, we'll just note that package.json exists
	// In a real implementation, you'd parse the JSON
	return []sdk.Dependency{
		{
			Name:          "npm-dependencies",
			Version:       "detected",
			Type:          "npm",
			ARMCompatible: "unknown",
		},
	}, nil
}

// scanPythonDependencies scans Python dependencies from requirements.txt
func (s *Scanner) scanPythonDependencies() ([]sdk.Dependency, error) {
	requirementsPath := filepath.Join(s.projectPath, "requirements.txt")
	file, err := os.Open(requirementsPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var dependencies []sdk.Dependency
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		// Parse package name and version
		var name, version string
		if strings.Contains(line, "==") {
			parts := strings.Split(line, "==")
			name = strings.TrimSpace(parts[0])
			if len(parts) > 1 {
				version = strings.TrimSpace(parts[1])
			}
		} else if strings.Contains(line, ">=") {
			parts := strings.Split(line, ">=")
			name = strings.TrimSpace(parts[0])
			if len(parts) > 1 {
				version = ">=" + strings.TrimSpace(parts[1])
			}
		} else {
			name = line
			version = "*"
		}

		dependencies = append(dependencies, sdk.Dependency{
			Name:          name,
			Version:       version,
			Type:          "python",
			ARMCompatible: "unknown",
		})
	}

	return dependencies, scanner.Err()
}

// scanCargoDependencies scans Rust dependencies from Cargo.toml
func (s *Scanner) scanCargoDependencies() ([]sdk.Dependency, error) {
	cargoTomlPath := filepath.Join(s.projectPath, "Cargo.toml")
	if _, err := os.Stat(cargoTomlPath); os.IsNotExist(err) {
		return nil, err
	}

	// For simplicity, we'll just note that Cargo.toml exists
	// In a real implementation, you'd parse the TOML
	return []sdk.Dependency{
		{
			Name:          "cargo-dependencies",
			Version:       "detected",
			Type:          "cargo",
			ARMCompatible: "unknown",
		},
	}, nil
}

// scanGoDependencies scans Go dependencies from go.mod
func (s *Scanner) scanGoDependencies() ([]sdk.Dependency, error) {
	goModPath := filepath.Join(s.projectPath, "go.mod")
	file, err := os.Open(goModPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var dependencies []sdk.Dependency
	scanner := bufio.NewScanner(file)
	inRequireBlock := false

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		
		if strings.HasPrefix(line, "require (") {
			inRequireBlock = true
			continue
		}
		
		if inRequireBlock && line == ")" {
			inRequireBlock = false
			continue
		}

		if inRequireBlock && line != "" && !strings.HasPrefix(line, "//") {
			parts := strings.Fields(line)
			if len(parts) >= 2 {
				name := parts[0]
				version := parts[1]

				dependencies = append(dependencies, sdk.Dependency{
					Name:          name,
					Version:       version,
					Type:          "go",
					ARMCompatible: "unknown",
				})
			}
		}
	}

	return dependencies, scanner.Err()
}

// generateRecommendations generates recommendations based on scan results
func (s *Scanner) generateRecommendations(results *sdk.ScanResults) []string {
	var recommendations []string

	issueCount := len(results.Issues)
	highSeverityCount := 0
	for _, issue := range results.Issues {
		if issue.Severity == "high" {
			highSeverityCount++
		}
	}

	if issueCount == 0 {
		recommendations = append(recommendations, "✅ No obvious x86-specific code detected")
	} else {
		recommendations = append(recommendations, fmt.Sprintf("🔍 Found %d potential compatibility issues", issueCount))
		
		if highSeverityCount > 0 {
			recommendations = append(recommendations, fmt.Sprintf("⚠️  %d high-severity issues require immediate attention", highSeverityCount))
		}
	}

	// Build system recommendations
	buildSystemTypes := make(map[string]bool)
	for _, bs := range results.BuildSystems {
		buildSystemTypes[bs.System] = true
	}

	if buildSystemTypes["cmake"] {
		recommendations = append(recommendations, "📋 CMake detected - review CMakeLists.txt for architecture-specific settings")
	}
	if buildSystemTypes["make"] {
		recommendations = append(recommendations, "📋 Makefile detected - review for architecture-specific compiler flags")
	}

	// Dependency recommendations
	depCount := len(results.Dependencies)
	if depCount > 0 {
		recommendations = append(recommendations, fmt.Sprintf("📦 %d dependencies found - verify ARM compatibility", depCount))
	}

	return recommendations
}

// isVerbose returns true if verbose mode is enabled
func (s *Scanner) isVerbose() bool {
	if verbose, ok := s.config["verbose"].(bool); ok {
		return verbose
	}
	return false
}

File: sdk/scanner.py
"""
Code scanner module for detecting x86-specific code and dependencies.
"""

import logging
import re
import ast
import subprocess
from typing import Dict, List, Any, Set
from pathlib import Path

logger = logging.getLogger(__name__)


class CodeScanner:
    """
    Scans code for x86-specific instructions, dependencies, and compatibility issues.
    """

    def __init__(self, project_path: Path, config: Dict[str, Any]):
        self.project_path = project_path
        self.config = config

        # Define patterns for x86-specific code
        self.x86_patterns = {
            "inline_assembly": [
                r"__asm__\s*\(",
                r"asm\s*\(",
                r"_asm\s*{",
            ],
            "x86_intrinsics": [
                r"#include\s*<.*mmintrin\.h.*>",
                r"#include\s*<.*xmmintrin\.h.*>",
                r"#include\s*<.*emmintrin\.h.*>",
                r"#include\s*<.*pmmintrin\.h.*>",
                r"#include\s*<.*immintrin\.h.*>",
                r"_mm_\w+",
                r"_mm\d+_\w+",
            ],
            "architecture_checks": [
                r"#ifdef\s+_M_X64",
                r"#ifdef\s+__x86_64__",
                r"#ifdef\s+_M_IX86",
                r"#ifdef\s+__i386__",
            ],
            "platform_specific": [
                r"GetSystemInfo",
                r"IsWow64Process",
                r"SYSTEM_INFO",
            ],
        }

        # File extensions to scan
        self.scannable_extensions = {
            ".c",
            ".cpp",
            ".cc",
            ".cxx",
            ".h",
            ".hpp",
            ".hxx",
            ".py",
            ".go",
            ".rs",
            ".java",
            ".cs",
            ".js",
            ".ts",
            ".jsx",
            ".tsx",
        }

    def scan(self) -> Dict[str, Any]:
        """
        Scan the entire project for x86-specific code.

        Returns:
            Dictionary containing scan results
        """
        results = {
            "total_files": 0,
            "scanned_files": 0,
            "issues": [],
            "dependencies": [],
            "build_systems": [],
            "recommendations": [],
        }

        logger.info(f"Scanning project at: {self.project_path}")

        # Scan source files
        for file_path in self._get_source_files():
            results["total_files"] += 1

            try:
                file_issues = self._scan_file(file_path)
                if file_issues:
                    results["issues"].extend(file_issues)
                results["scanned_files"] += 1

            except Exception as e:
                logger.warning(f"Failed to scan {file_path}: {str(e)}")

        # Scan build systems
        results["build_systems"] = self._scan_build_systems()

        # Scan dependencies
        results["dependencies"] = self._scan_dependencies()

        # Generate recommendations
        results["recommendations"] = self._generate_recommendations(results)

        logger.info(
            f"Scan completed: {results['scanned_files']}/{results['total_files']} files"
        )
        return results

    def _get_source_files(self) -> List[Path]:
        """Get list of source files to scan."""
        source_files = []

        for file_path in self.project_path.rglob("*"):
            if (
                file_path.is_file()
                and file_path.suffix.lower() in self.scannable_extensions
                and not self._should_skip_file(file_path)
            ):
                source_files.append(file_path)

        return source_files

    def _should_skip_file(self, file_path: Path) -> bool:
        """Check if file should be skipped during scanning."""
        skip_patterns = [
            ".git",
            "__pycache__",
            "node_modules",
            ".venv",
            "venv",
            "build",
            "dist",
            ".tox",
            ".pytest_cache",
        ]

        return any(pattern in str(file_path) for pattern in skip_patterns)

    def _scan_file(self, file_path: Path) -> List[Dict[str, Any]]:
        """Scan a single file for x86-specific code."""
        issues = []

        try:
            with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
                content = f.read()

            # Check for x86-specific patterns
            for category, patterns in self.x86_patterns.items():
                for pattern in patterns:
                    matches = list(
                        re.finditer(pattern, content, re.IGNORECASE | re.MULTILINE)
                    )

                    for match in matches:
                        line_num = content[: match.start()].count("\n") + 1

                        issue = {
                            "file": str(file_path.relative_to(self.project_path)),
                            "line": line_num,
                            "category": category,
                            "pattern": pattern,
                            "matched_text": match.group(0),
                            "severity": self._get_severity(category),
                            "suggestion": self._get_suggestion(
                                category, match.group(0)
                            ),
                        }
                        issues.append(issue)

        except Exception as e:
            logger.warning(f"Error scanning file {file_path}: {str(e)}")

        return issues

    def _get_severity(self, category: str) -> str:
        """Get severity level for issue category."""
        severity_map = {
            "inline_assembly": "high",
            "x86_intrinsics": "high",
            "architecture_checks": "medium",
            "platform_specific": "medium",
        }
        return severity_map.get(category, "low")

    def _get_suggestion(self, category: str, matched_text: str) -> str:
        """Get suggestion for fixing the issue."""
        suggestions = {
            "inline_assembly": "Replace with portable C/C++ code or use ARM NEON intrinsics",
            "x86_intrinsics": "Replace with ARM NEON equivalents or portable alternatives",
            "architecture_checks": "Add ARM architecture checks or use runtime detection",
            "platform_specific": "Use cross-platform alternatives or add ARM-specific implementations",
        }
        return suggestions.get(category, "Review for ARM compatibility")

    def _scan_build_systems(self) -> List[Dict[str, Any]]:
        """Scan for build system configurations."""
        build_files = []

        build_patterns = {
            "CMakeLists.txt": "cmake",
            "Makefile": "make",
            "*.mk": "make",
            "build.gradle": "gradle",
            "pom.xml": "maven",
            "package.json": "npm",
            "Cargo.toml": "cargo",
            "go.mod": "go_modules",
            "*.pro": "qmake",
        }

        for pattern, build_system in build_patterns.items():
            matches = list(self.project_path.glob(pattern))
            matches.extend(list(self.project_path.rglob(pattern)))

            for match in matches:
                build_files.append(
                    {
                        "file": str(match.relative_to(self.project_path)),
                        "system": build_system,
                        "needs_review": True,
                    }
                )

        return build_files

    def _scan_dependencies(self) -> List[Dict[str, Any]]:
        """Scan project dependencies for ARM compatibility."""
        dependencies = []

        # Check package.json (Node.js)
        package_json = self.project_path / "package.json"
        if package_json.exists():
            dependencies.extend(self._scan_npm_dependencies(package_json))

        # Check requirements.txt (Python)
        requirements_txt = self.project_path / "requirements.txt"
        if requirements_txt.exists():
            dependencies.extend(self._scan_python_dependencies(requirements_txt))

        # Check Cargo.toml (Rust)
        cargo_toml = self.project_path / "Cargo.toml"
        if cargo_toml.exists():
            dependencies.extend(self._scan_cargo_dependencies(cargo_toml))

        # Check go.mod (Go)
        go_mod = self.project_path / "go.mod"
        if go_mod.exists():
            dependencies.extend(self._scan_go_dependencies(go_mod))

        return dependencies

    def _scan_npm_dependencies(self, package_json: Path) -> List[Dict[str, Any]]:
        """Scan npm dependencies."""
        dependencies = []

        try:
            import json

            with open(package_json, "r") as f:
                data = json.load(f)

            all_deps = {}
            all_deps.update(data.get("dependencies", {}))
            all_deps.update(data.get("devDependencies", {}))

            for name, version in all_deps.items():
                dependencies.append(
                    {
                        "name": name,
                        "version": version,
                        "type": "npm",
                        "arm_compatible": "unknown",  # Would need to check npm registry
                    }
                )

        except Exception as e:
            logger.warning(f"Error scanning package.json: {str(e)}")

        return dependencies

    def _scan_python_dependencies(self, requirements_txt: Path) -> List[Dict[str, Any]]:
        """Scan Python dependencies."""
        dependencies = []

        try:
            with open(requirements_txt, "r") as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith("#"):
                        # Parse package name and version
                        if "==" in line:
                            name, version = line.split("==", 1)
                        elif ">=" in line:
                            name, version = line.split(">=", 1)
                        else:
                            name, version = line, "*"

                        dependencies.append(
                            {
                                "name": name.strip(),
                                "version": version.strip(),
                                "type": "python",
                                "arm_compatible": "unknown",  # Would need to check PyPI
                            }
                        )

        except Exception as e:
            logger.warning(f"Error scanning requirements.txt: {str(e)}")

        return dependencies

    def _scan_cargo_dependencies(self, cargo_toml: Path) -> List[Dict[str, Any]]:
        """Scan Rust dependencies."""
        dependencies = []

        try:
            import toml

            with open(cargo_toml, "r") as f:
                data = toml.load(f)

            deps = data.get("dependencies", {})
            for name, config in deps.items():
                if isinstance(config, str):
                    version = config
                else:
                    version = config.get("version", "*")

                dependencies.append(
                    {
                        "name": name,
                        "version": version,
                        "type": "cargo",
                        "arm_compatible": "unknown",  # Would need to check crates.io
                    }
                )

        except Exception as e:
            logger.warning(f"Error scanning Cargo.toml: {str(e)}")

        return dependencies

    def _scan_go_dependencies(self, go_mod: Path) -> List[Dict[str, Any]]:
        """Scan Go dependencies."""
        dependencies = []

        try:
            with open(go_mod, "r") as f:
                content = f.read()

            # Parse require blocks
            require_pattern = r"require\s+\((.*?)\)"
            require_blocks = re.findall(require_pattern, content, re.DOTALL)

            for block in require_blocks:
                lines = block.strip().split("\n")
                for line in lines:
                    line = line.strip()
                    if line and not line.startswith("//"):
                        parts = line.split()
                        if len(parts) >= 2:
                            name = parts[0]
                            version = parts[1]

                            dependencies.append(
                                {
                                    "name": name,
                                    "version": version,
                                    "type": "go",
                                    "arm_compatible": "unknown",  # Would need to check
                                }
                            )

        except Exception as e:
            logger.warning(f"Error scanning go.mod: {str(e)}")

        return dependencies

    def _generate_recommendations(self, results: Dict[str, Any]) -> List[str]:
        """Generate recommendations based on scan results."""
        recommendations = []

        issue_count = len(results["issues"])
        high_severity_count = sum(
            1 for issue in results["issues"] if issue["severity"] == "high"
        )

        if issue_count == 0:
            recommendations.append("✅ No obvious x86-specific code detected")
        else:
            recommendations.append(
                f"🔍 Found {issue_count} potential compatibility issues"
            )

            if high_severity_count > 0:
                recommendations.append(
                    f"⚠️  {high_severity_count} high-severity issues require immediate attention"
                )

        # Build system recommendations
        build_systems = [bs["system"] for bs in results["build_systems"]]
        if "cmake" in build_systems:
            recommendations.append(
                "📋 CMake detected - review CMakeLists.txt for architecture-specific settings"
            )
        if "make" in build_systems:
            recommendations.append(
                "📋 Makefile detected - review for architecture-specific compiler flags"
            )

        # Dependency recommendations
        dep_count = len(results["dependencies"])
        if dep_count > 0:
            recommendations.append(
                f"📦 {dep_count} dependencies found - verify ARM compatibility"
            )

        return recommendations
